helix-term/src/handlers/signature_help.rs

```diff
@@ -184,54 +210,50 @@ pub fn show_signature_help(
     let doc = doc!(editor);
     let language = doc.language_name().unwrap_or("");
 
-    let signature = match response
+    if response.signatures.is_empty() {
+        return;
+    }
+
+    let signatures: Vec<Signature> = response
         .signatures
-        .get(response.active_signature.unwrap_or(0) as usize)
-    {
-        Some(s) => s,
-        None => return,
-    };
-    let mut contents = SignatureHelp::new(
-        signature.label.clone(),
+        .into_iter()
+        .map(|s| {
+            let active_param_range = active_param_range(&s, response.active_parameter);
+
+            let signature_doc = if config.lsp.display_signature_help_docs {
+                s.documentation.map(|doc| match doc {
+                    lsp::Documentation::String(s) => s,
+                    lsp::Documentation::MarkupContent(markup) => markup.value,
+                })
+            } else {
+                None
+            };
+
+            Signature {
+                signature: s.label,
+                signature_doc,
+                active_param_range,
+            }
+        })
+        .collect();
+
+    let old_popup = compositor.find_id::<Popup<SignatureHelp>>(SignatureHelp::ID);
+    let mut active_signature = old_popup
+        .as_ref()
+        .map(|popup| popup.contents().active_signature())
+        .unwrap_or_else(|| response.active_signature.unwrap_or_default() as usize);
+
+    if active_signature >= signatures.len() {
+        active_signature = signatures.len() - 1;
+    }
```

Comment:
I think that we should prefer the `active_signature` we get from the language server here instead:

```rust
let active_signature = response
    .active_signature
    .map(|n| n as usize)
    .or_else(|| {
        old_popup
            .as_ref()
            .map(|popup| popup.contents().active_signature())
    })
    .unwrap_or_default()
    .min(signatures.len() - 1);
```

Some language servers update the `active_signature` they send when you enter enough parameters. For example with Erlang you might be typing `maps:get(` which is a multiarity function. `maps:get(a, b` should show the signature of `maps:get/2` (takes 2 arguments) but for `maps:get(a, b,` the language server will show `maps:get/3`.

We could make this more complicated though and separately track which signature the language server sends and which you navigate to with A-n/A-p, preferring the signature you navigate to until the server changes which signature it considers active.

Reply:
I have played with your code. My c++ LSP somehow does update the active signature at all. it always `0`
so IMO it is very annoying to switch to the right signature, then it switches always back  to the first one.

but in the same time using the C# LSP your code brings some improvements, since the suggested LSP signature is in most case useful.

I created #10655, which implements the bit more complicated solution you have suggested